### PR TITLE
Render void elements as self-closing tags

### PR DESCRIPTION
**Pull request by jonathantneal**
## Changes

- Fixes an issue where void elements are rendered with opening and closing tags.
- Resolves https://github.com/withastro/astro/issues/2062

## Testing

no tests added

## Docs

bug fix only

### CODE REVIEW

**Comment by matthewp** [comment]
Can this just be `>` and not `/>`?

File: packages/astro/src/runtime/server/index.ts [hunk 1]
```diff
@@ -223,7 +225,7 @@ If you're still stuck, please open an issue on GitHub or join us at https://astr
   // This is a custom element without a renderer. Because of that, render it
   // as a string and the user is responsible for adding a script tag for the component definition.
   if (!html && typeof Component === 'string') {
-    html = await renderAstroComponent(await render`<${Component}${spreadAttributes(props)}>${children}</${Component}>`);
+    html = await renderAstroComponent(await render`<${Component}${spreadAttributes(props)}${voidElementNames.test(Component) ? `/>` : `>${children}</${Component}>`}`);
```
Comment (jonathantneal) [inline packages/astro/src/runtime/server/index.ts:228]: > Can this just be `>` and not `/>`?

Yes, it could. My preference is to be explicit.

**Comment by drwpow** [comment]
Could we add a few tests for this, maybe in `astro-basic`? It would be good to ensure `>` vs `/>` doesn’t affect it. It’d also be good to test for things like not closing SVG `<path>` elements (not that our compiler should transform them; just that they don’t break and render as-is)

**Comment by jonathantneal** [comment]
Some tests of the capability have been added, @drwpow.

File: packages/astro/test/astro-basic.test.js [hunk 2]
```diff
@@ -83,6 +83,26 @@ describe('Astro basics', () => {
...
 
+  it('Supports void elements whose name is a string (#2062)', async () => {
+    const html = await fixture.readFile('/input/index.html');
+    const $ = cheerio.load(html);
+
+    // <Input />
+    expect($('body > :nth-child(1)').prop('outerHTML')).to.equal('<input>');
+
+    // <Input type="password" />
+    expect($('body > :nth-child(2)').prop('outerHTML')).to.equal('<input type="password">');
```
Comment (drwpow) [inline packages/astro/test/astro-basic.test.js:94]: Awesome tests!

**Review by drwpow — APPROVED** [review]
Great tests, thank you! And good fix

File: packages/astro/src/runtime/server/index.ts [hunk 3]
```diff
@@ -223,7 +225,9 @@ If you're still stuck, please open an issue on GitHub or join us at https://astr
   // This is a custom element without a renderer. Because of that, render it
   // as a string and the user is responsible for adding a script tag for the component definition.
   if (!html && typeof Component === 'string') {
-    html = await renderAstroComponent(await render`<${Component}${spreadAttributes(props)}>${children}</${Component}>`);
+    html = await renderAstroComponent(
+      await render`<${Component}${spreadAttributes(props)}${(children == null || children == '') && voidElementNames.test(Component) ? `/>` : `>${children}</${Component}>`}`
```
Comment (jonathantneal) [inline packages/astro/src/runtime/server/index.ts:229]: Eh, `==`?

Yea, it’s that or `children === undefined || children === null || String(children) === ''`.

Comment (drwpow) [reply packages/astro/src/runtime/server/index.ts:229]: I think this is good enough. I can’t think of a problem with `==` here

**Pull request merged**